studies/tasks.py

```diff
@@ -187,12 +189,51 @@ def _segmented_by_study(validated_groups):
         yield user, dict(map_reduce(child_study_pairs, itemgetter(1), itemgetter(0)))
 
 
-def acquire_announcement_email_targets() -> Generator:
+def acquire_potential_announcement_email_targets() -> Generator:
     return _segmented_by_study(
         _validated(_deserialized(_grouped_by_user(potential_message_targets())))
     )
 
 
+def limit_email_targets(
+    potential_targets_segmented_by_study, max_emails_per_study
+) -> Generator:
+    """Reduces number of targets so users get <= 1 email and studies get <= max_emails_per_study.
+
+    This is done AFTER deserializing to actual model objects so that we've already checked eligibility;
+    otherwise we'd have to select a random sample of families to MAYBE email if they turn out to be
+    eligible each time. """
+
+    from studies.models import Study
+
+    # Iterate through generator with actual models to build a list of IDs
+    all_user_study_children_tuples = []
+    for user, study_child_mapping in potential_targets_segmented_by_study:
+        study, child_list = random.choice(list(study_child_mapping.items()))
+        all_user_study_children_tuples.append(
+            (user.id, study.id, [child.id for child in child_list])
+        )
+
+    # Randomly select the first <= N study-user pairs for each study. We don't want to just yield the first N per study
+    # because then we'll always invite some families to participate first, others later
+    random.shuffle(all_user_study_children_tuples)
```

Comment:
What about just doing `potential_targets_as_list = list(potential_targets_segmented_by_study)` to build the list in memory, then `random.shuffle(potential_targets_as_list)` and iterating over that? No need to throw the objects away and re-fetch them from postgres, especially since we're limited to 100 parents from the upstream generator (`_deserialized`)

Reply:
Ohh, I must have misunderstood - I thought if we iterated over the entire `_deserialized` output we would end up getting through all the potential targets, not 100 at a time - and that this was part of why we'd want to avoid generating such a list. That's what we want, so that we don't always start with the same 100, then the next 100, etc. - shuffling within those potential targets wouldn't really help.

If this would only make a list of 100 users, how come we were able to send substantially more than 100 emails when iterating over the output in `send_announcement_emails`?

Reply:
Sorry, that was sloppy phrasing: a maximum of 100 grouped family-to-study mappings are present in the upstream `_deserialized` generator at any time - furthermore, these groupings are "hydrated" from the results of a batched query to avoid shooting off a bazillion queries. Since it's in a generator pipeline, you still only effectively hydrate just one group at a time before it gets checked in `_validate`. 

The reason for mentioning that is just to say that in building out the full list, there's an upper bound on how many objects are getting allocated on the heap upstream. You'll still get the full list of hydrated targets when you exhaust the generator with `list(potential_targets_segmented_by_study)`

Reply:
Ohh I see - ok, so building the list of validated targets doesn't involve ever building the whole big list they come from. I'd still be concerned that, say, someday someone wants to run a study that's for ages 0-18 and the list of validated targets is large, or that there are several new studies sending out email across a variety of ages. I.e., if the big list might choke the µwsgi worker, I'd be nervous making assumptions that the smaller list never will.

Reply:
We do use `more_itertools` - if you wanted to get real fancy pantsy you could use the [`sample` function](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.sample), which uses [optimized reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling#An_optimal_algorithm) under the hood. The tradeoff there is that you're not guaranteeing an explicit # maximum emails per study - just picking a random sample of families and hoping that you're covering all the studies fairly

Reply:
Yeah, that was my initial hope but the issue is that you end up getting roughly exponential decay in emails sent per study at some point, instead of actually just emailing everyone eligible on a reasonable timescale. This matters a little more for studies with narrow inclusion criteria - e.g. if you want to email everyone with a newborn, or everyone with a kid with ASD - which are "competing" with studies with broader criteria. Ideally we want to actually exhaust all the eligible kids in a simple-to-understand way. So I think we do actually want to go up to a maximum per study rather than sampling families, unless you're concerned about any of the costs involved here.